### PR TITLE
Storage Write API Batch mode - All Sink Task methods

### PR DESCRIPTION
**Pull request by b-goyal**
This PR adds all SInkTask related changes for batch mode in Storage Write API and updates the corresponding units tests
JIRA: CC-20071
EPIC: CC-19644

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java
Comment (SubhamSinghal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:88]: we should allow other parameters of JsonStreamWriter to configurable e.g.
```
ignoreUnknownValues
customExecutorProvider
enableConnectionMultiplexing
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java
Comment (SubhamSinghal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: response.get() would block the thread. Should we make it return `CompletableFuture` instead?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:88]: Thanks for the inputs @SubhamSinghal . Currently connector does not expose these configs. Making JsonStreamWriter configurable would mean adding adding these configs as user-configurable that is currently not in scope for first iteration. Though we can look into this and prioritise in future iterations.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: Th execution of this part is already delegated to a thread pool. This would not block the main thread.

Comment (SubhamSinghal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: but this would block thread in the threadpool. Would we be able to use this under heavy load? I am thinking under heavy traffic, this would cause frequent flow control because threads may be busy waiting for response.

Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: Yes. We are yet to do performance testing but the primary reasons for choosing this approach is Non-blocking call handling did not fit in well with retry mechanism and exception handling design we had. This has been tested on average load and works fine. For special cases where load is high, there are configs which can be tweaked to keep the connector performant.  Also we have batch mode which should cover for it.

Comment (SubhamSinghal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: We have usecase where we load high traffic of data through storage write api. We don't want to opt for batch load as it does not provide SLA gurantee. Can we look into resiliency4J library for async retry?

Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java:147]: Thanks for the inputs @SubhamSinghal. We will do adequate performance testing on our end to verify the limits. 
Also, we will consider looking into the suggested library.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (aniketshrimal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:625]: What is the reason behind having a value of 1 here?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (aniketshrimal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:626]: Any reason we are going with value 10 for initial delay and period?

**Review by aniketshrimal — COMMENTED** [review]
LGTM overall @sp-gupta! Have some minor comment.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:210]: Sure, will move.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:625]: We want to only have 1 thread committing streams.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:626]: This is part of test block. Real block takes user input.

Comment (aniketshrimal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:626]: Just curious - where is the test block being called, i.e, what cases do we have testStorageWriteApi not null? Also is it not possible to just update the config values to 10 instead of having a separate block since these values are configurable.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQueryStorageApiBatchSinkTaskTest.java
Comment (aniketshrimal) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQueryStorageApiBatchSinkTaskTest.java:137]: I am not sure when this code will be triggered. Is it not the case that it will throw error on previous line and this while loop won't get executed - `testTask.put(Collections.singletonList(spoofSinkRecord()));`?

**Review by aniketshrimal — COMMENTED** [review]
LGTM overall! Approving with some minor comments.

**Review by pbadani — APPROVED** [review]

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQueryStorageApiBatchSinkTaskTest.java
Comment (b-goyal) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQueryStorageApiBatchSinkTaskTest.java:137]: No actually. We execute threads in put method and line:127 here would raise an exception in one of the threads. We rely on this method ([maybeThrowEncounteredError](https://github.com/confluentinc/kafka-connect-bigquery/blob/691dddc5a1f9224df40588b9fbe239cd88b63775/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L301) ) to see if any exception occurred in previous calls to put and fail task. Since the thread execution is async we cannot confidently say in which call to put will there be exception so the while loop. 
line:129 would not throw any exception.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (b-goyal) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:626]: We have a different constructor for testing [ref](https://github.com/confluentinc/kafka-connect-bigquery/pull/312/files#diff-ad5086c1123b437f762b169166c133be1b56528bd22eba426ec7dea10110f9b2R157) . Only when created via that constructor testStorageWriteApi would be not null. Config range starts from 1 minute. 1-2 minutes felt too long for unit test and 10 seconds if passed via config would have failed.

**Comment by b-goyal** [comment]
@pbadani  and I had a discussion on refactoring BigQuerySinkTask needs . CC-20419 is the tracking ticket for same. This is not in scope for current iteration and would be taken up in next iterations.

**Pull request merged**
